rdm/audit_for_gaps.py

```diff
@@ -175,7 +178,69 @@ def _sorted_checklist_items(unsorted_checklist):
         key = checklist_item.get('reference', '')
         description = checklist_item.get('description', '')
         unsorted_items.append(key + ' ' + description)
-    return sorted(unsorted_items)
+    # The use of the SectionalAnalysis tool ensures '62304:5.1.2' < '62304:5.1.11' < '62304:5.2.1'
+    return sorted(unsorted_items, key=SectionalAnalysis)
+
+
+class SectionalAnalysis:
+    def __init__(self, text):
+        self.components = _components(text)
+
+    def __lt__(self, other):
+        my_size = len(self.components)
+        other_size = len(other.components)
+        min_size = min(my_size, other_size)
+        # Use first difference...
+        for index in range(min_size):
+            my_number, my_text = self.components[index]
+            other_number, other_text = other.components[index]
+            if my_number < other_number:
+                return True
+            elif my_number > other_number:
+                return False
+            if my_text < other_text:
+                return True
+            elif my_text > other_text:
+                return False
+        # ... or else shorter is lesser
+        return my_size < other_size
+
+
+def _components(text):
+    if text:
+        number, non_number, remainder = _next_component(text)
```

Comment:
@orwonthe are we assuming that only the last item can be a letter?

Reply:
Short answer, no.
The text is broken into a sequence of number, text pairs. The case where the head of the text is non-numeric is handled as well as the case where the tail of the text is numeric. The sequences produced by two items are compared in order. The comparisons alternate between numeric and non-numeric until a difference is found. The first comparison is numeric, the second is textual, the third is numeric,... until a difference is found.